Datagram Splitting

dr7ana opened the pull request:
Datagram splitting PR, implementation details in commit messages

dr7ana committed:
Packet splitting support

- refactored DatagramIO::pending_datagrams to return a one-shot usable struct for ngtcp2
- 007 test cases for packet splitting
- removed lazy splitting (nonsensical)

tewinget commented on tests/007-datagrams.cpp:
```diff
@@ -160,87 +151,115 @@ namespace oxen::quic::test
 
-            std::promise<bool> tls_promise, data_promise;
-            std::future<bool> tls_future = tls_promise.get_future(), data_future = data_promise.get_future();
+            while (data.size() < 2000)
```
This number may as well reflect max datagram size.

dr7ana replied:
good thought, will reflect in the updated 007 test when i add the rotating buffer test

tewinget commented:
Since we're only splitting in half, there should be a test which attempts to send a packet which won't fit in 2 datagrams and requires that `send_datagram` throws.

tewinget commented:
We'll also want tests to make sure the rotating buffer behavior is correct.  That should be interesting to design positive and negative tests for, and of course then implement.

dr7ana committed:
Datagram splitting support

- split datagrams sent when needed
- configurable rotating buffer in datagramIO
- unit tests for basic coverage on rotating buffer and datagram splitting (to add more later)

tewinget commented on tests/007-datagrams.cpp:
```diff
@@ -279,15 +298,317 @@ namespace oxen::quic::test
+
+            std::atomic<int> index{0};
+            std::atomic<int> data_counter{0};
+            size_t bufsize = 256, n = bufsize / 2 + 1;
```
not a huge fan of multiple variable declarations on the same line.  also since `n` is used in several places and is not like a simple counter,  it could use a more descriptive name.

tewinget commented on include/quic/opt.hpp:
```diff
@@ -43,10 +43,16 @@ namespace oxen::quic::opt
     {
         bool split_packets = false;
         Splitting mode = Splitting::NONE;
+        int bufsize = 4096;
```
should probably be a `size_t` or some other unsigned type.

jagerman commented on include/quic/datagram.hpp:
```diff
@@ -1,16 +1,71 @@
+    struct prepared_datagram
+    {
+        uint64_t id;                  // internal ID for ngtcp2
+        std::array<uint8_t, 2> dgid;  // optional transmitted ID buffer (for packet splitting)
```
Shouldn't this be just a uint16_t?

jagerman replied:
Wait, nevermind, I see how this is used.

tewinget commented on tests/007-datagrams.cpp:
```diff
@@ -279,15 +298,317 @@ namespace oxen::quic::test
+
+            auto server_ci = server_endpoint->get_all_conns(Direction::INBOUND).front();
+
+            REQUIRE(server_ci->last_cleared() == 0);
```
This confirms that `last_cleared` was set to the expected value, but it doesn't confirm that the bucket was actually cleared.  In order to do that, we would need to have a way to make the connection send one half of a packet but not the other half, and then we'd need to send enough packets to loop back around do that packet number and make sure the other side of the connection gets the correct new packet rather than some frankensteined packet.

dr7ana replied:
keep scrolling, theres a test case that does exactly that

jagerman commented on include/quic/datagram.hpp:
```diff
@@ -1,16 +1,71 @@
+        uint16_t id{0};
+        bstring data{};
+        // -1 = first half, 1 = second half
+        int part{0};
```
Any reason not just to use a bool (maybe `first_part`) for this?

Also, whether bool or not, this would be slightly better placed between `id` and `data` to make the struct layout *slightly* smaller: `bstring` most likely uses size_t-alignment (i.e. 8-byte alignment), so if you do it this way you have a struct containing data like this: `[u16][6 bytes padding][string, probably 24 bytes][int][4 bytes padding]`, whereas if you rearrange the smaller types together you will get: `[u16][2 padding][int][24-byte-string]` and make the struct 8 bytes smaller.  Not a huge deal, but also easy to change.

dr7ana replied:
I used an int so I can reset it to 0 when the entry is cleared

Good point about the alignment, will fix it

jagerman commented on include/quic/datagram.hpp:
```diff
@@ -1,16 +1,71 @@
 
 namespace oxen::quic
 {
+    extern std::atomic<bool> enable_test_features;
```
Perhaps rename to `debug_drop_datagrams`?

jagerman commented on include/quic/datagram.hpp:
```diff
@@ -1,16 +1,71 @@
 namespace oxen::quic
 {
+    extern std::atomic<bool> enable_test_features;
+    extern std::atomic<int> test_counter;
```
These two should also be inside `#ifndef NDEBUG`

jagerman commented on tests/007-datagrams.cpp:
```diff
@@ -279,15 +298,317 @@ namespace oxen::quic::test
+    {
+        if (disable_rotating_buffer)
+            SKIP("Rotating buffer testing not enabled for this test iteration!");
+
```
```suggestion

#ifdef NDEBUG
        SKIP("Induced loss test requires a debug build");
#else
```

jagerman replied:
and `#endif` down at the bottom of the section

dr7ana replied:
added!

jagerman commented on src/connection.cpp:
```diff
@@ -740,14 +865,41 @@ namespace oxen::quic
-        // append received datagram to tetris_buffer if packet_splitting is enabled
-        if (_packet_splitting)
+        std::optional<bstring> maybe_data;
+        uint16_t dgid = oxenc::load_big_to_host<uint16_t>(data.data());
```
When packet splitting is completely disabled on the connection we ~~shouldn't be~~ *aren't* sending a dgid, so need to not try to read one here.

dr7ana replied:
fixed

jagerman commented on src/connection.cpp:
```diff
@@ -758,7 +910,7 @@ namespace oxen::quic
             try
             {
-                datagrams->dgram_data_cb(data);
+                datagrams->dgram_data_cb((maybe_data ? *maybe_data : bstring{data.begin(), data.end()}));
```
```suggestion
                datagrams->dgram_data_cb((maybe_data ? std::move(*maybe_data) : bstring{data.begin(), data.end()}));
```

dr7ana replied:
fixed!

jagerman commented on src/connection.cpp:
```diff
@@ -740,14 +865,41 @@ namespace oxen::quic
+            catch (std::exception& e)
+            {
+                log::error(log_cat, "Exception: {}", e.what());
+                return -1;
```
Should this be NGTCP2_ERR_CALLBACK_FAILURE?

dr7ana replied:
that's a good question... on one hand it doesn't matter because any non-zero return value buffs out to `NGTCP2_ERR_CALLBACK_FAILURE` [upon return](https://github.com/ngtcp2/ngtcp2/blob/d80f9b3d709da79cdbd2b94acfae3073810de80d/lib/ngtcp2_conn.c#L589-L593) so we might as well return that... on the other hand that specific line isn't a failure in the callback, but in the buffer.

maybe we could extend std exception and make our own that we throw and handle internally?

jagerman commented on src/datagram.cpp:
```diff
@@ -3,73 +3,257 @@
+        log::trace(log_cat, "{} called", __PRETTY_FUNCTION__);
+
+        {
+            std::lock_guard<std::mutex> buffer_lock(m);
```
I *think* we can drop the mutex and atomics in here, if we ensure that we only ever touch them from the main thread.  (For the test suite I think that would mean we need to add some indirection that goes through a `call` rather than accessing things directly, but that's probably fine for the test suite).

jagerman commented on src/datagram.cpp:
```diff
@@ -3,73 +3,257 @@
 
-            temp.base = const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(b.first.data()));
-            temp.len = b.first.size();
+        auto r = recv_buffer.receive(data, dgid);
```
I think we can just `return recv_buffer.receive(data, dgid);` here?

jagerman commented on src/datagram.cpp:
```diff
@@ -3,73 +3,257 @@
+
+                bool order = (b.part < 0);  // if true, we have the first part already stored
+                // must use substring or we will end up with two datagram ID's at the beginning and middle
+                out.append((order ? b.data.substr(2) : data.substr(2)));
+                out.append((order ? data.substr(2) : b.data.substr(2)));
```
Unfortunately, string.substr allocates and copies into a brand new string.  You could avoid it by wrapping in a string_view first, like this:
```suggestion
                out.append(bstring_view{order ? b.data : data}.substr(2));
                out.append(bstring_view{order ? data : b.data}.substr(2));
```

jagerman replied:
Perhaps better would be to do the substring earlier, so that we don't keep the dgid in the data bstring at all

dr7ana replied:
yeah that's a good point, didn't think about that! i will do what your later comment suggests and drop the prefix earlier

jagerman commented on src/datagram.cpp:
```diff
@@ -3,73 +3,257 @@
+                log::trace(log_cat, "Pairing datagram (ID: {}) with counterpart at buffer pos [{},{}]", dgid, row, col);
+
+                bstring out{};
+                out.reserve(data.size() + b.data.size());
```
I think that we could potentially avoid another string allocation here by reusing either `b.data` or `data`.  Maybe something like:

```C++
bstring out;
if (order) {
    b.data.append(data);
    out = std::move(b.data);
} else {
    data.append(b.data);
    out = std::move(data);
}
```

Then we could *also* reserve space when we first set up the partial `.data` element so that this append won't need to reallocate.

jagerman replied:
No, wait, the second version here is wrong.

jagerman replied:
v2:
```C++
if (order)
    b.data.append(data);
else {
    b.data.reserve(b.data.size() + data.size());
    // We have the second half and need to prepend the first half, so move it to the end:
    std::memmove(b.data.data() + data.size(), b.data.data(), b.data.size());
    // Now copy the first half into place:
    std::memcpy(b.data.data(), data.data(), data.size());
}
bstring out{std::move(b.data)};
```
which should need no new allocation *if* we pre-reserved the needed amount when creating the data.

jagerman replied:
Also those three lines have a lot of Data.  Lore will be upset.

dr7ana replied:
quick note about the memmoves: when we already have the first half of the datagram, the incoming smaller half is overwriting the big half. this is with the new storage structures, so ill adjust and let you know

jagerman replied:
The `reserve` should be a `resize` instead.

jagerman commented on src/datagram.cpp:
```diff
@@ -3,73 +3,257 @@
+
+            log::trace(log_cat, "Storing datagram (ID: {}) at buffer pos [{},{}]", dgid, row, col);
+
+            b = received_datagram{dgid, data};
```
... that is, just before this line we can do `data.remove_prefix(2);` to discard the dgid from the data we store.

dr7ana replied:
or even earlier in connection?

jagerman commented on include/quic/datagram.hpp:
```diff
@@ -1,16 +1,71 @@
+        int part{0};
+
+        received_datagram() = default;
+        explicit received_datagram(uint16_t dgid, bstring_view d) : id{dgid}, data{d} { part = (dgid % 4 == 2) ? -1 : 1; };
```
We know that *at most* the assembled datagram will be `d.size() + MAX_DATAGRAM_SIZE` (this is probably the wrong name for that constant), so we can reserve it with:
```suggestion
        explicit received_datagram(uint16_t dgid, bstring_view d) : id{dgid}, part{(dgid % 4 == 2) ? -1 : 1} {
            data.reserve(d.size() + MAX_DATAGRAM_SIZE);
            data.append(d);
        }
```
and then we fill out this buffer with the remaining packet when it arrives without having to copy and allocate a new buffer (left a comment over there about this).

dr7ana replied:
awesome

dr7ana committed:
Flip flop datagram sending implemented

- datagrams can be sent preferentially with the big or small portion first
- refactored send buffer into a more logical deq
- added 007 test case for flip flopping messages
- dynamic message sizing in flip flop test case just for mac CI

jagerman commented on src/datagram.cpp:
```diff
@@ -3,73 +3,275 @@
+            else
+            {
+                b.data.reserve(b.data.size() + data.size());
+                b.data.insert(0, data.data());
```
Ah yeah, that's cleaner than my version.

jagerman committed:
Add datagrams atomically

dr7ana committed:
Datagram test fixups

- debug test counters are now static Connection member variables
- tests can count on server and client side
- moved messaging containers to messages.{hpp,cpp}

tewinget commented on include/quic/connection.hpp:
```diff
@@ -307,6 +264,13 @@ namespace oxen::quic
         // returns number of currently pending streams for use in test cases
         size_t num_pending() const { return pending_streams.size(); }
+
+#ifndef NDEBUG
```
supplanted by `debug_interface`

tewinget commented on include/quic/connection.hpp:
```diff
@@ -41,7 +41,22 @@ namespace oxen::quic
+    class debug_interface
+    {
+      public:
+        std::atomic<bool> enable_datagram_drop_test;
```
reads like a verb

dr7ana replied:
just do it

jagerman commented on include/quic/connection.hpp:
```diff
@@ -71,17 +86,21 @@ namespace oxen::quic
         }
 
-        // TOFIX: We can't template virtual functions; do this a better way if it ends up working
+        // note: We can't template virtual functions; do this a better way if it ends up working
```
You can make this work by having a `protected` base pure virtual method (maybe called `send_impl` or something like that), and then templatizing as much as you want here in `connection_interface` where each templated version recasts its arguments and passes them off to `send_impl` (so that an implementing class need only implement/override the single `send_impl` method).

dr7ana replied:
resolving this after our mumble chat

jagerman commented on include/quic/connection.hpp:
```diff
@@ -142,7 +159,11 @@ namespace oxen::quic
+        // public debug functions; to be removed with friend test fixture class
+        int datagrams_stored() override;  // TOFIX: this one
+        int last_cleared() override;
+        int datagram_bufsize() override;
```
~~To remove?~~

jagerman replied:
Never mind, these are (potentially) useful for an application that maybe wants to worry about how packets are arriving.

Perhaps add a one-line comment above each to give a quick description of what they are?

dr7ana replied:
you got it

tewinget commented on include/quic/stream.hpp:
```diff
@@ -219,6 +223,12 @@ namespace oxen::quic
             }
         };
 
+        prepared_datagram pending_datagram(std::atomic<bool>&) override
```
This should probably throw?

dr7ana replied:
good idea

jagerman commented on include/quic/datagram.hpp:
```diff
@@ -1,16 +1,26 @@
+    // extern std::atomic<int> test_drop_counter;
+    // extern std::atomic<bool> enable_datagram_flip_flop_test;
+    // extern std::atomic<int> test_flip_flop_counter;
+#endif
```
Delete

jagerman commented on include/quic/datagram.hpp:
```diff
@@ -32,16 +41,16 @@ namespace oxen::quic
+        virtual void send(bstring_view, std::shared_ptr<void> keep_alive = nullptr) = 0;
         virtual std::vector<ngtcp2_vec> pending() = 0;
-        virtual void append_buffer(bstring_view buffer, std::shared_ptr<void> keep_alive) = 0;
+        virtual prepared_datagram pending_datagram(std::atomic<bool>&) = 0;
```
Not clear why this needs to be atomic

tewinget commented on src/connection.cpp:
```diff
@@ -467,9 +583,15 @@ namespace oxen::quic
+
+                // if we just loaded stream frames into the packet, we should not reverse the order of the next split
+                // datagram
+                // prefer_big_first = false;
```
if this is dead code, should be removed

jagerman replied:
~~Should this be uncommented?~~

Dead code.

tewinget commented on src/connection.cpp:
```diff
@@ -417,6 +531,8 @@ namespace oxen::quic
         pkt_tx_timer_updater pkt_updater{*this, ts};
         size_t stream_packets = 0;
 
+        std::atomic<bool> prefer_big_first{true};
```
shouldn't need to be atomic

jagerman replied:
Since only `flush_packets` touches this (and flush_packets will only ever be called inside the main thread) this ought to be safe as a plain `bool`

tewinget commented on src/connection.cpp:
```diff
@@ -478,15 +600,19 @@ namespace oxen::quic
+                {
+                    log::trace(log_cat, "ngtcp2 accepted datagram ID: {} for transmission", dgram.id);
+                    datagrams->send_buffer.drop_front(prefer_big_first);
+                    // datagrams->send_buffer.erase(dgram.target);
```
dead code

jagerman replied:
Dead code

jagerman commented on include/quic/messages.hpp:
```diff
@@ -0,0 +1,142 @@
+        {
+            data.reserve(d.size() + MAX_PMTUD_UDP_PAYLOAD);
+            data.append(d);
+        };
```
```suggestion
        }
```

jagerman commented on include/quic/messages.hpp:
```diff
@@ -0,0 +1,142 @@
+            id = 0;
+            part = 0;
+            data.clear();
+        };
```
```suggestion
        }
```

jagerman commented on include/quic/connection.hpp:
```diff
@@ -71,17 +86,21 @@ namespace oxen::quic
+        // public debug functions: remove this with testfixture friend classes
+        virtual int datagrams_stored() = 0;
+        virtual int last_cleared() = 0;
+        virtual int datagram_bufsize() = 0;
```
I think these can all be `const`

jagerman commented on include/quic/datagram.hpp:
```diff
@@ -84,14 +172,30 @@ namespace oxen::quic
+    - more rows can increase tolerance in misordering
+    - too many short ones leads to pinstriping
+
+*/
```
Leftover notes?

jagerman commented on include/quic/datagram.hpp:
```diff
@@ -84,14 +172,30 @@ namespace oxen::quic
+        std::vector<ngtcp2_vec> pending() override
+        {
+            log::warning(log_cat, "{} called", __PRETTY_FUNCTION__);
+            return std::vector<ngtcp2_vec>{};
```
Purely a style thing (i.e. feel completely free to ignore), but this can be written as `return {};` if you don't feel like repeating the return type again.

dr7ana replied:
i appreciate it thank you

jagerman commented on include/quic/stream.hpp:
```diff
@@ -219,6 +223,12 @@ namespace oxen::quic
+        {
+            log::warning(log_cat, "{} called", __PRETTY_FUNCTION__);
+            return prepared_datagram{};
+        };
```
```suggestion
        }
```

jagerman commented on src/connection.cpp:
```diff
@@ -191,15 +217,106 @@ namespace oxen::quic
+        {
+            log::trace(log_cat, "Note: CID-{} in closing period; signaling endpoint to delete connection", scid());
+
+            _endpoint.call([&]() {
```
Another completely ignorable style point: for a lambda with no arguments you can leave out the `()` entirely, i.e.
```suggestion
            _endpoint.call([&] {
```

dr7ana replied:
noted and fixed

jagerman commented on src/connection.cpp:
```diff
@@ -294,8 +411,6 @@ namespace oxen::quic
     };
 
-    int64_t sent_counter = 0;
-
```
Whoops, my bad here leaving that in in the udp refactor.

dr7ana replied:
ill re-add it

jagerman commented on src/datagram.cpp:
```diff
@@ -3,73 +3,112 @@
+// std::atomic<bool> oxen::quic::enable_datagram_flip_flop_test;
+// std::atomic<int> oxen::quic::test_drop_counter;
+// std::atomic<int> oxen::quic::test_flip_flop_counter;
+#endif
```
dead code

jagerman commented on src/datagram.cpp:
```diff
@@ -3,73 +3,112 @@
+        //         else
+        //         {
+        //             log::debug(log_cat, "enable_datagram_flip_flop_test is false, skipping optional logic");
+        //         }
```
dead code

tewinget commented on include/quic/messages.hpp:
```diff
@@ -0,0 +1,142 @@
+    struct buffer_que
+    {
+        std::deque<datagram_storage> buf{};
+        size_t quantity{0};
```
appears unused

tewinget commented on include/quic/messages.hpp:
```diff
@@ -0,0 +1,142 @@
+        size_t quantity{0};
+
+        bool empty() const { return buf.empty(); }
+        size_t size() const { return quantity; }
```
should return `buf.size()` ?

jagerman commented on src/connection.cpp:
```diff
@@ -191,15 +217,106 @@ namespace oxen::quic
+        {
+            log::trace(log_cat, "Note: CID-{} in closing period; signaling endpoint to delete connection", scid());
+
+            _endpoint.call([&]() {
```
Here it might be cleaner/tighter to capture `[this]` rather than `[&]` as it looks like the current object is the only thing we care about capturing inside the lambda.

dr7ana replied:
good point thank you

tewinget commented on src/connection.cpp:
```diff
@@ -191,15 +217,106 @@ namespace oxen::quic
+            log::debug(log_cat, "Error: connection is already draining; dropping");
+        }
+
+        // TODO: if read packet gives us failure, should we close?
```
This appears to be covered within `read_packet`

dr7ana replied:
good catch thats an old comment

jagerman commented on src/connection.cpp:
```diff
@@ -191,15 +217,106 @@ namespace oxen::quic
+                break;
+            case NGTCP2_ERR_DRAINING:
+                log::trace(log_cat, "Note: CID-{} is draining; signaling endpoint to drain connection", scid());
+                _endpoint.call([&]() {
```
These (this one and the various ones that follow) could also be `[this]` captures

dr7ana committed:
Datagram splitting misc review fixes
